jormungandr/src/explorer/mod.rs

```diff
@@ -267,54 +263,46 @@ impl ExplorerDB {
                             stake_pool_blocks,
                             &explorer_block,
                         ),
-                    ))
+                    )
                 }
-                None => Err(Error::from(ErrorKind::AncestorNotFound(format!(
-                    "{}",
-                    block.id()
-                )))),
-            })
-            .and_then(
-                move |(
-                    parent_ref,
+                None => {
+                    return Err(Error::from(ErrorKind::AncestorNotFound(format!(
+                        "{}",
+                        block.id()
+                    ))))
+                }
+            };
+
+        let chain_length = chain_length.clone();
+        let (stake_pool_data, stake_pool_blocks) = stake_pools;
+
+        let state_ref = multiverse
+            .insert(
+                chain_length,
+                block_id,
+                State {
+                    parent_ref: Some(parent_ref),
                     transactions,
                     blocks,
                     addresses,
                     epochs,
                     chain_lengths,
-                    stake_pools,
-                )| {
-                    let chain_length = chain_length.clone();
-                    let block_id = block_id.clone();
-                    let (stake_pool_data, stake_pool_blocks) = stake_pools;
-                    multiverse
-                        .insert(
-                            chain_length,
-                            block_id,
-                            State {
-                                parent_ref: Some(parent_ref),
-                                transactions,
-                                blocks,
-                                addresses,
-                                epochs,
-                                chain_lengths,
-                                stake_pool_data,
-                                stake_pool_blocks,
-                            },
-                        )
-                        .map_err(|_: Infallible| unreachable!())
-                        .map(move |state_ref| (state_ref, chain_length))
+                    stake_pool_data,
+                    stake_pool_blocks,
                 },
             )
-            .and_then(move |(state_ref, chain_length)| {
-                current_tip
-                    .compare_and_replace(Branch {
-                        state_ref: state_ref.clone(),
-                        length: chain_length,
-                    })
-                    .map_err(|_: Infallible| unreachable!())
-                    .map(|_| state_ref)
+            .await;
+
+        current_tip
+            .compare_and_replace(Branch {
+                state_ref: state_ref.clone(),
+                length: chain_length,
             })
+            .compat()
```

Comment:
Not converted yet, but can wait for another PR.

Reply:
Will continue working on that, but yes probably should do that in a separate PR.